### PR TITLE
WDPD-177 Ratepay fraud protection

### PR DESCRIPTION
**Pull request by ps-cgrach**
### This PR

Add special fraud protection handling for Ratepay.

### How to Test

- Before going to the checkout page, open the developer console.
- Check the network to load a `di.js` with a unique token in front in the path.
- Do a purchase.
- Check the request `device - fingerprint` should be the same unique token.

### Jira Ticket
https://jira.parkside.at/browse/WDPD-177

### CODE REVIEW

File: Core/Helper.php [hunk 1]
```diff
@@ -9,18 +9,15 @@
 
 namespace Wirecard\Oxid\Core;
 
+use DateTime;
+use Exception;
```
Comment (nickstamat) [inline Core/Helper.php]: [minor] Maybe add a newline to separate from the following group.

Comment (ps-holler) [reply Core/Helper.php]: Currently we only enforce an empty line if there is a base path and it changes from one line to another, for example:

```php
use Wirecard\Oxid\Model\PaymentMethod;
use Wirecard\Oxid\Model\Transaction;

use OxidEsales\Eshop\Core\Registry;
use OxidEsales\Eshop\Core\Exception\StandardException;

use DateTime;
use Exception;
```

This is also implemented in the PHPCS ruleset.

File: Core/Helper.php [hunk 2]
```diff
@@ -366,4 +363,19 @@ public static function getShop()
...
     }
+
+    /**
+     * Generates a unique token
+     *
+     * @return string
+     *
+     * @since 1.2.0
+     */
+    public static function getUniqueToken()
```
Comment (ps-holler) [inline Core/Helper.php]: `OxidEsales\EshopCommunity\Core\UniversallyUniqueIdGenerator` has some methods to create unique tokens built in.

File: views/blocks/profiling_tags.tpl [hunk 3]
```diff
@@ -11,6 +11,26 @@
 
 [{if $oPayment->isCustomPaymentMethod()}]
 
+  [{if $oPayment->oxpayments__oxid->value === 'wdratepay-invoice'}]
+
+  [{assign var="sDeviceIdentToken" value=$oViewConf->getRatepayUniqueToken()}]
+  <script language="JavaScript">
```
Comment (ps-holler) [inline views/blocks/profiling_tags.tpl]: I guess this is just copied over, but the `language` attribute has been deprecated for a long time and should not be used.

File: views/blocks/profiling_tags.tpl [hunk 4]
```diff
@@ -11,6 +11,26 @@
...
+    var di = {t:"[{$sDeviceIdentToken}]",v:"WDWL",l:"Checkout"};
+  </script>
+  <script type="text/javascript" src="//d.ratepay.com/[{$sDeviceIdentToken}]/di.js">
+  </script>
+  <noscript>
+    <link rel="stylesheet" type="text/css"
+          href="//d.ratepay.com/di.css?t=[{$sDeviceIdentToken}]&v=WDWL&l=Checkout">
+  </noscript>
+  <object type="application/x-shockwave-flash" data="//d.ratepay.com/[{$sDeviceIdentToken}]/c.swf" width="0"
```
Comment (ps-holler) [inline views/blocks/profiling_tags.tpl]: I think we can safely ignore this Flash fallback, even for OXID. Most modern browsers stopped supporting Flash long ago and [Adobe is going to finally pull the plug next year](https://theblog.adobe.com/adobe-flash-update/).

File: Core/Helper.php (hunk 2 above)
Comment (ps-cgrach) [reply Core/Helper.php]: that token is for ratepay and has to have the session id

File: views/blocks/profiling_tags.tpl (hunk 3 above)
Comment (ps-cgrach) [reply views/blocks/profiling_tags.tpl]: @tomazpu should I delete the language attribute?

File: Core/Helper.php (hunk 1 above)
Comment (ps-cgrach) [reply Core/Helper.php]: 👍

File: views/blocks/profiling_tags.tpl (hunk 4 above)
Comment (ps-cgrach) [reply views/blocks/profiling_tags.tpl]: @tomazpu Any objections?

File: views/blocks/profiling_tags.tpl (hunk 3 above)
Comment (tomazpu) [reply views/blocks/profiling_tags.tpl]: yes

File: views/blocks/profiling_tags.tpl (hunk 4 above)
Comment (tomazpu) [reply views/blocks/profiling_tags.tpl]: I understand, but as it is officially documented with this fallback please leave it in.

Comment (ps-cgrach) [reply views/blocks/profiling_tags.tpl]: 👍

File: views/blocks/profiling_tags.tpl (hunk 3 above)
Comment (ps-cgrach) [reply views/blocks/profiling_tags.tpl]: 👍

File: Core/Helper.php [hunk 5]
```diff
@@ -368,6 +366,7 @@ public static function getShop()
     }
 
     /**
+<<<<<<< HEAD
```
Comment (tomazpu) [inline Core/Helper.php]: I think this is a leftover from a merge conflict

Comment (ps-cgrach) [reply Core/Helper.php]: removed 😓

**Pull request merged**